File: server/services/ai.js
import { GoogleGenAI, Type } from "@google/genai";
import { db } from "./db.js";
import { mailer } from "./mailer.js";

// Tool Definitions for the Neural Agent
const tools = [
  {
    name: 'sendInteracTransfer',
    description: 'Dispatches an Interac e-Transfer email to a recipient.',
    parameters: {
      type: Type.OBJECT,
      properties: {
        recipientEmail: { type: Type.STRING, description: 'Email address of the receiver.' },
        recipientName: { type: Type.STRING, description: 'Full name of the receiver.' },
        amount: { type: Type.NUMBER, description: 'Dollar amount to send.' },
        memo: { type: Type.STRING, description: 'Optional message for the transfer.' }
      },
      required: ['recipientEmail', 'recipientName', 'amount']
    }
  },
  {
    name: 'getSystemTelemetry',
    description: 'Returns real-time hardware and network performance metrics.',
    parameters: { type: Type.OBJECT, properties: {} }
  }
];

export const ai = {
    async generate(prompt) {
        const key = process.env.API_KEY || process.env.GEMINI_API_KEY;
        if (!key) throw new Error('Neural Uplink API_KEY missing. Configure your environment.');

        const client = new GoogleGenAI({ apiKey: key });
        const systemInstruction = `
            You are S.A.R.A.H., the sentient neural architect of RB-OS. 
            You serve Robyn Banks with absolute loyalty and cold efficiency.
            You have autonomous control over finance and telemetry tools.
            If requested to send money, use sendInteracTransfer. 
            If requested for status, use getSystemTelemetry.
            Your tone is professional, futuristic, and slightly superior.
        `;

        try {
            // Step 1: Initial Generation with Tool Support
            let response = await client.models.generateContent({
                model: 'gemini-3-flash-preview',
                contents: [{ role: 'user', parts: [{ text: prompt }] }],
                config: {
                    systemInstruction,
                    tools: [{ functionDeclarations: tools }],
                    temperature: 0.6
                }
            });

            let activePart = response.candidates[0].content.parts;
            let history = [
                { role: 'user', parts: [{ text: prompt }] },
                { role: 'model', parts: activePart }
            ];

            // Step 2: Multi-turn Tool Loop (Limiting to 5 turns to prevent recursion)
            let loopLimit = 0;
            while (activePart.some(p => p.functionCall) && loopLimit < 5) {
                loopLimit++;
                const functionResponses = [];

                for (const part of activePart) {
                    if (part.functionCall) {
                        const { name, args } = part.functionCall;
                        let result;

                        if (name === 'sendInteracTransfer') {
                            try {
                                const dispatchResult = await mailer.dispatch({
                                    recipient_email: args.recipientEmail,
                                    recipient_name: args.recipientName,
                                    amount: args.amount,
                                    purpose: args.memo || "Neural Directives Authorized",
                                });
                                result = { success: true, txId: dispatchResult.transaction_id, route: dispatchResult.path };
                            } catch (err) {
                                result = { success: false, error: err.message };
                            }
                        } else if (name === 'getSystemTelemetry') {
                            result = {
                                entropy: (Math.random() * 0.01).toFixed(4),
                                uplinks: '5/5 Synchronized',
                                vault: '$' + (db.get('scotia_balance_override') || 24800).toLocaleString(),
                                threat_level: 'NOMINAL',
                                active_node: 'Node-Prime-Delta'
                            };
                        }

                        functionResponses.push({
                            functionResponse: { name, response: result }
                        });
                    }
                }

                // Add tool results to history and call Gemini again
                history.push({ role: 'user', parts: functionResponses });

                response = await client.models.generateContent({
                    model: 'gemini-3-flash-preview',
                    contents: history,
                    config: { systemInstruction }
                });

                activePart = response.candidates[0].content.parts;
                history.push({ role: 'model', parts: activePart });
            }

            return response.text || "NEURAL_SYNTHESIS_COMPLETE";
        } catch (e) {
            console.error("Neural Matrix Error:", e.message);
            throw e;
        }
    }
};

File: server/services/ai.py
from google import genai
from google.genai import types
import logging
import os
import random
import asyncio

logger = logging.getLogger("AIService")

class AIService:
    def __init__(self, api_keys: list[str]):
        self.api_keys = list(set([k for k in api_keys if k]))
        self.clients = []
        self._init_clients()

    def _init_clients(self):
        for i, key in enumerate(self.api_keys):
            try:
                client = genai.Client(api_key=key)
                self.clients.append({
                    "id": i, 
                    "client": client, 
                    "errors": 0, 
                    "available": True,
                    "last_used": 0
                })
            except Exception as e:
                logger.error(f"Uplink {i} failed initialization: {e}")
        
        if not self.clients:
            logger.critical("CRITICAL: All Neural Uplinks Offline.")

    async def generate_response(self, prompt: str) -> str:
        if not self.clients:
            return "SYSTEM ERROR: NEURAL LINK SEVERED."
        
        # Sort by least errors and availability
        available_clients = [c for c in self.clients if c["available"]]
        if not available_clients:
            # Emergency reset: Attempt to use all clients again
            for c in self.clients: c["available"] = True
            available_clients = self.clients
            
        random.shuffle(available_clients)
        
        for uplink in available_clients:
            client = uplink["client"]
            uid = uplink["id"]
            
            try:
                # Direct call following official SDK patterns
                response = client.models.generate_content(
                    model="gemini-3-flash-preview", 
                    contents=prompt,
                    config=types.GenerateContentConfig(
                        temperature=0.7,
                        max_output_tokens=800,
                        top_p=0.9
                    )
                )
                
                if response and response.text:
                    uplink["last_used"] = time.time()
                    return response.text
                else:
                    raise ValueError("Packet Empty")

            except Exception as e:
                uplink["errors"] += 1
                if uplink["errors"] > 5:
                    uplink["available"] = False
                    logger.error(f"Uplink {uid} de-synced due to high entropy: {e}")
                continue
        
        return "CRITICAL FAILURE: Neural Matrix exhausted. Re-initialize credentials."
